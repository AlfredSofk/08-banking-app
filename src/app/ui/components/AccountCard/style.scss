

.account-card {
    background-color: #ffffff;
    padding: 1.5rem;
    border-radius: 8px;
    box-shadow: 0px 4px 12px rgba(0, 0, 0, 0.1);
    width: 220px;
    text-align: center;
    transition: transform 0.3s ease, box-shadow 0.3s ease;
  
    &:hover {
      transform: translateY(-5px);
      box-shadow: 0px 6px 18px rgba(0, 0, 0, 0.15);
    }
  
    &__title {
      font-size: 1.2rem;
      font-weight: bold;
      color: #333;
      margin-bottom: 0.5rem;
    }
  
    &__number {
      font-size: 0.9rem;
      color: #777;
      margin-bottom: 1rem;
    }
  
    &__balance {
      font-size: 1.1rem;
      color: #1abc9c;
      font-weight: bold;
    }
  }
  