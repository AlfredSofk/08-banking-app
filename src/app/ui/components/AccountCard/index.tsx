

import { decryptAES } from '../../../core/utils/decoder'
import './style.scss'

interface Props {
    accountNumber : string,
    balance: number
}

export const AccountCard = ({accountNumber, balance} : Props) => {
    return (
        <div className="account-card">
            {/* <h3 className="account-card__title">{accountName}</h3> */}
            <p className="account-card__number" data-testid="accountNumber">NÃºmero de Cuenta: {accountNumber}</p>
            <p className="account-card__balance" data-testid="balance">Saldo: ${balance.toFixed(2)}</p>
        </div>
    )
}