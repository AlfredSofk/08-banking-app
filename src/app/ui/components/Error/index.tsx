import React from "react";
import "./style.scss";

interface ErrorMessageProps {
  title?: string; // Título del mensaje de error
  description?: string; // Descripción del error
  errorDetails?: string; // Detalles adicionales del error
}

export const ErrorScreen: React.FC<ErrorMessageProps> = ({
  title = "Something went wrong",
  description = "An unexpected error occurred. Please try again later.",
  errorDetails,
}) => {
  return (
    <section
      className="error-message"
      role="alert"
      aria-live="assertive"
      aria-atomic="true"
    >
      <header className="error-message__header">
        <h1 className="error-message__title">{title}</h1>
      </header>
      <div className="error-message__body">
        <p className="error-message__description">{description}</p>
        {errorDetails && (
          <details className="error-message__details">
            <summary className="error-message__summary">More details</summary>
            <p>{errorDetails}</p>
          </details>
        )}
      </div>
    </section>
  );
};