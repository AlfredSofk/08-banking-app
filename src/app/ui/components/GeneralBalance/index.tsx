import { FaDollarSign } from "react-icons/fa";
import './style.scss';

export const GeneralBalance = ({ balance }: { balance: number }) => {

    return (
        <div className="general-balance">
        <h3 className="general-balance__title">Balance General</h3>
        <div className="general-balance__content">
          <FaDollarSign aria-hidden="true" className="general-balance__amount"/>
          <p className="general-balance__amount" role="general-balance">{balance.toFixed(2)}</p>
        </div>
      </div>
    )

}