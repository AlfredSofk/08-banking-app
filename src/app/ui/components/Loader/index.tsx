import './style.scss'

export const Loader = () => {
  return (
    <div className="loader" aria-label='Cargando...' data-testid='loader'></div>
  );
};