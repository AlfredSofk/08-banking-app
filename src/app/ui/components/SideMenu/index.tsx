import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { FaHome, FaUser, FaCog, FaQuestion, FaBars, FaChevronDown } from 'react-icons/fa';
import { CiBank } from 'react-icons/ci';
import './style.scss';

export const SideMenu = () => {
  const [isExpanded, setIsExpanded] = useState(true);
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
  const navigate = useNavigate();

  const handleToggleSidebar = () => {
    setIsExpanded(!isExpanded);
  };

  const handleNavigate = (path: string) => {
    console.log(path)
    navigate(path);
  };

  return (
    <aside className={`sidebar ${isExpanded ? 'sidebar--expanded' : ''}`} aria-label="Barra lateral de navegación">
      <button className="sidebar__toggle" onClick={handleToggleSidebar} aria-label="Expandir menú">
        <FaBars />
      </button>
      <div className="sidebar__logo"><CiBank /><p>BankU</p></div>

      <ul className="sidebar__menu">
        <li className="sidebar__item">
          <button
            className="sidebar__button"
            aria-label="Inicio"
            onClick={() => handleNavigate('/home/inicio')}
          >
            <FaHome />
            {isExpanded && <span className="sidebar__text">Inicio</span>}
          </button>
        </li>

        <li className="sidebar__item sidebar__item--dropdown">
          <button
            className="sidebar__button"
            onClick={() => setIsDropdownOpen(!isDropdownOpen)}
            aria-haspopup="true"
            aria-expanded={isDropdownOpen}
          >
            <FaUser />
            {isExpanded && <span className="sidebar__text">Transacciones</span>}
            {isExpanded && <FaChevronDown className="sidebar__icon--dropdown" />}
          </button>
          {isDropdownOpen && isExpanded && (
            <ul className="sidebar__submenu">
              <li className="sidebar__submenu-item">
                <button className="sidebar__submenu-button" onClick={() => handleNavigate('/home/withdraw')}>
                  Retirar Cajero
                </button>
              </li>
              <li className="sidebar__submenu-item">
                <button className="sidebar__submenu-button" onClick={() => handleNavigate('/home/depositATM')}>
                  Depositar Cajero
                </button>
              </li>
              <li className="sidebar__submenu-item">
                <button className="sidebar__submenu-button" onClick={() => handleNavigate('/home/deposit-account')}>
                  Depósito
                </button>
              </li>
              <li className="sidebar__submenu-item">
                <button className="sidebar__submenu-button" onClick={() => handleNavigate('/home/transfer')}>
                  Transferencias
                </button>
              </li>
              <li className="sidebar__submenu-item">
                <button className="sidebar__submenu-button" onClick={() => handleNavigate('/home/purchase-web')}>
                  Compra Web
                </button>
              </li>
              <li className="sidebar__submenu-item">
                <button className="sidebar__submenu-button" onClick={() => handleNavigate('/home/purchase-store')}>
                  Compra Establecimiento
                </button>
              </li>
            </ul>
          )}
        </li>

        <li className="sidebar__item">
          <button
            className="sidebar__button"
            aria-label="Cuenta"
            onClick={() => handleNavigate('/home/account')}
          >
            <FaCog />
            {isExpanded && <span className="sidebar__text">Cuenta</span>}
          </button>
        </li>

        <li className="sidebar__item">
          <button
            className="sidebar__button"
            aria-label="Ayuda"
            onClick={() => handleNavigate('/home/about')}
          >
            <FaQuestion />
            {isExpanded && <span className="sidebar__text">Acerca de</span>}
          </button>
        </li>
      </ul>
    </aside>
  );
};
