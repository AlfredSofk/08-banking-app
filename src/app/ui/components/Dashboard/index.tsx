import { IAppState } from "../../../core/interfaces/bankAccount";
import { AccountCard } from "../AccountCard";
import { GeneralBalance } from "../GeneralBalance";
import './style.scss'

interface Props { 
    state : IAppState
}


export const Dashboard = ({state} : Props) => {

    const {userAccounts, saldoGlobal} = state

    return (
        <section className="accounts-overview" aria-labelledby="overview-heading">
        <h2 id="overview-heading" className="accounts-overview__title" role="heading">Resumen de Cuentas</h2>
        <GeneralBalance balance={saldoGlobal} />
        <div className="accounts-overview__accounts">
          {userAccounts.length === 0 && <p>No hay cuentas disponibles</p>}
          {userAccounts.map((account, index) => (
            <AccountCard
              key={index}
            //   accountName={get}
              accountNumber={account.number}
              balance={account.amount}
            />
          ))}
        </div>
      </section>
    )

}