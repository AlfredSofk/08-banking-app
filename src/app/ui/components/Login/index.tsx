// import { useLoginToken } from '../../../core/hooks/useLoginToken';
// import { USER_LOGIN } from '../../../core/constants/user';
// import { useContext } from 'react';
// import { AuthContext } from '../../../core/state/authContext/AuthContext';
import './style.scss';
import { IAuthState } from '../../../core/interfaces/auth';
import { IBodyLoginToken } from '../../../core/interfaces/requestToApi';
import { SubmitHandler, useForm } from 'react-hook-form';
import { ILoginToken } from '../../../core/interfaces/requestApiTo';
import { Loader } from '../Loader';


interface Props {
    loginUser:  (data: IBodyLoginToken) => void
    state: IAuthState
}

interface IFormInput {
    usuario: string
    password: string
}

export default function LoginForm({state,loginUser}: Props) {

    const { register, watch,handleSubmit, formState : {errors}} = useForm<IFormInput>({
        defaultValues : {
            usuario : '',
            password : ''
        }
    }) 

    const onSubmit : SubmitHandler<IFormInput> = (data) => {

        let dataForm : IBodyLoginToken = {
            username : data.usuario,
            password : data.password
        }
        console.log(data)
        loginUser(dataForm)

    }
    return (
        <>
            <div className="login">
                <section className='login__container' aria-labelledby="login-heading">
                    <h1 id="login-heading" className='login__title'>Iniciar Sesión</h1>
                    
                    <form className='login-form' aria-describedby="login-form-desc" onSubmit={handleSubmit(onSubmit)}>
                        <p id="login-form-desc" className='login-form__desc'>
                            Introduce tus credenciales para acceder a tu cuenta.
                        </p>
                        <div className='login-form__group'>
                            <label htmlFor="usuario" className='login-form__label'>Usuario</label>
                            <input 
                            type="usuario" 
                            id="usuario" 
                            className='login-form__input' 
                            placeholder="usuarioEjemplo" 
                            aria-required="true"
                            aria-describedby="usuario-desc"
                            aria-invalid={errors.usuario ? "true" : "false"}
                            {...register("usuario", { required: true, pattern: /^[a-zA-Z0-9]+$/i})}
                            />
                            {errors.usuario && (<small className='login-form__error' role="alert">El campo usuario es requerido</small>)}
                            <span id="usuario-desc" className="sr-only">Introduce un correo válido</span>
                        </div>

                        <div className='login-form__group'>
                            <label htmlFor="password" className='login-form__label'>Contraseña</label>
                            <input 
                            type="password" 
                            id="password" 
                            className='login-form__input' 
                            placeholder="••••••••" 
                            aria-required="true"
                            aria-describedby="password-desc"
                            aria-invalid={errors.password ? "true" : "false"}
                            {...register("password", { required: true, minLength: 8 })}                        
                            />
                            {errors.password && <small className='login-form__error' role="alert">El campo password es requerido</small>}
                            <span id="password-desc" className="sr-only">Introduce tu contraseña</span>
                        </div>
                        {
                        state.loading 
                            ?<Loader/>
                            :<button type='submit'  className='login-form__button' >Iniciar Sesión</button>
                        }
                        
                    </form>

                    <p className='login__register'>
                    ¿No tienes una cuenta? 
                    {/* <button type="button" className='login__register-link' aria-label="Registrarse">Regístrate aquí</button> */}
                    </p>
                </section>
            </div>

        </>
    )

}