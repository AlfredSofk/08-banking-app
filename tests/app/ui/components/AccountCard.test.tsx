import { render, screen } from "@testing-library/react";
import { AccountCard } from "../../../../src/app/ui/components/AccountCard";
import React from "react";



describe('test para el componente AccountCard', () => {


    test('match snapshot', () => {

        const props = {
            accountNumber: "123456789", 
            balance :1000
        }

        const { asFragment } = render( 
            <AccountCard {...props} />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test('renderiza correctamente el número de cuenta y el saldo', () => {
        // Datos de prueba
        const props = {
            accountNumber: '1234567890',
            balance: 2500.5,
        };

        // Renderizar el componente
        render(<AccountCard {...props} />);

        // Verificar que el número de cuenta aparece correctamente
        expect(screen.getByTestId('accountNumber')).toBeInTheDocument();
        expect(screen.getByTestId('balance')).toBeInTheDocument();
        
        const accountNumber = screen.getByTestId('accountNumber');
        expect(accountNumber.textContent).toContain('Número de Cuenta: 1234567890');

        const balance = screen.getByTestId('balance');
        expect(balance.textContent).toContain('Saldo: $2500.50');

    });

    test('formatea correctamente el saldo', () => {
        // Datos de prueba con un saldo específico
        const props = {
            accountNumber: '9876543210',
            balance: 1234.567,
        };

        // Renderizar el componente
        render(<AccountCard {...props} />);

        // Verificar que el saldo está formateado a dos decimales
        expect(screen.getByText((content) =>
            content.includes(`$${props.balance.toFixed(2)}`)
        )).toBeInTheDocument();
    });

    test('muestra mensajes específicos si el número de cuenta o saldo están vacíos', () => {
        // Datos de prueba con valores vacíos
        const props = {
            accountNumber: '',
            balance: 0,
        };

        // Renderizar el componente
        render(<AccountCard {...props} />);

        // Verificar que aparece un mensaje predeterminado para el número de cuenta vacío
        expect(screen.getByText((content) => content.includes('Número de Cuenta:'))).toBeInTheDocument();

        // Verificar que el saldo se muestra como $0.00
        const balance = screen.getByTestId('balance');
        expect(balance.textContent).toContain('Saldo: $0.00');
    });

});