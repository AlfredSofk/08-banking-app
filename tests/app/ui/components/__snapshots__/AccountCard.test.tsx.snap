// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test para el componente AccountCard > match snapshot 1`] = `
<DocumentFragment>
  <div
    class="account-card"
    data-testid="account-123456789"
  >
    <p
      class="account-card__number"
      data-testid="accountNumber"
    >
      NÃºmero de Cuenta: 123456789
    </p>
    <p
      class="account-card__balance"
      data-testid="balance"
    >
      Saldo: $1000.00
    </p>
  </div>
</DocumentFragment>
`;
